src/cli/download_tracker.rs

```diff
@@ -1,291 +1,87 @@
         match *n {
             Notification::Install(In::Utils(Un::DownloadContentLengthReceived(content_len))) => {
                 self.content_length_received(content_len);
-
                 true
             }
             Notification::Install(In::Utils(Un::DownloadDataReceived(data))) => {
-                if self.stdout_is_a_tty {
-                    self.data_received(data.len());
-                }
+                self.data_received(data.len());
                 true
             }
             Notification::Install(In::Utils(Un::DownloadFinished)) => {
                 self.download_finished();
                 true
             }
-            Notification::Install(In::Utils(Un::DownloadPushUnit(unit))) => {
-                self.push_unit(unit);
-                true
-            }
-            Notification::Install(In::Utils(Un::DownloadPopUnit)) => {
-                self.pop_unit();
-                true
-            }
+            Notification::Install(In::Utils(Un::DownloadPushUnit(_))) => true,
+            Notification::Install(In::Utils(Un::DownloadPopUnit)) => true,
 
             _ => false,
         }
     }
 
-    /// Notifies self that Content-Length information has been received.
+    /// Sets the length for a new ProgressBar and gives it a style.
     pub(crate) fn content_length_received(&mut self, content_len: u64) {
-        self.content_len = Some(content_len as usize);
+        self.progress_bar.set_length(content_len);
+        self.progress_bar.set_style(
+            ProgressStyle::with_template(
+                "[{bar:40}] {bytes}/{total_bytes} ({bytes_per_sec}, ETA: {eta})",
+            )
+            .unwrap()
+            .progress_chars("## "),
+        );
     }
 
     /// Notifies self that data of size `len` has been received.
     pub(crate) fn data_received(&mut self, len: usize) {
-        self.total_downloaded += len;
-        self.downloaded_this_sec += len;
-
-        let current_time = Instant::now();
-
-        match self.last_sec {
-            None => self.last_sec = Some(current_time),
-            Some(prev) => {
-                let elapsed = current_time.saturating_duration_since(prev);
-                if elapsed >= Duration::from_secs(1) {
-                    if self.display_progress {
-                        self.display();
-                    }
-                    self.last_sec = Some(current_time);
-                    if self.downloaded_last_few_secs.len() == DOWNLOAD_TRACK_COUNT {
-                        self.downloaded_last_few_secs.pop_back();
-                    }
-                    self.downloaded_last_few_secs
-                        .push_front(self.downloaded_this_sec);
-                    self.downloaded_this_sec = 0;
-                }
-            }
+        if self.progress_bar.is_hidden() && self.progress_bar.elapsed() >= Duration::from_secs(1) {
+            self.multi_progress_bars.add(self.progress_bar.clone());
```

Comment:
Why do this?

Reply:
@djc This is to be coherent with the old rustup logic where if a progress is considered to finish instantly, don't actually show the related progress bar.